### PR TITLE
Bug fixes to get VQE working

### PR DESCRIPTION
**Pull request by bettinaheim**
This PR fixes the following issues in the QIR emission:
- processing callables that take a single named argument of a tuple type
- un-terminated blocks causing an exception for callables that return a value (blocks without predecessors or lack of termination due to improper handling of fail)
- empty callables not being defined
- wrong signature for function corresponding to controlled and controlled adjoint specializations
- generating wrapper functions (needed e.g. for partial applications) for functions that are part of the quantum instruction set
- using user defined types in partial applications
- tuple-valued arguments in partial applications, inlined callables, global callables, and callable values in all possible places (i.e. all kinds of call-like expressions)
- construction of the argument tuple for invoking controlled and controlled adjoint specializations for all kinds of call-like expressions
- handling of user defined types in callable values
- arguments of type unit in inlined callables
- assignments of allocated qubit tuples

As part of addressing these issues, this PR 
- makes sure all callables are built using the same logic and relying only on type information (opposed to user choices in the source code)
- unifies the construction of controlled arguments for all kinds of call-like expressions
- introduces common helpers for constructing and destructing tuples
- removes the special casing of tuple handling and argument construction/destruction in various calls
- processes assignments upon inlining in the same way as other assignments
- Uses IStructType opposed to the more general ITypeRef when a struct is expected

Since among the things that were not working properly were things that were used in multiple places (like for any one of the five distinctions made for calls), I unified the logic somewhat such that I only needed to fix them in one place. However, as a consequence of that, some of the generated IR looks quite redundant. I believe striving to get rid of that redundancy by introducing a lot of special casing in the emission is not the path to go, since a) it is a maintenance nightmare, b) subsequent compiler passes on the IR should eliminate that redundancy and if we wanted to make an effort to make the emitted IR more readable we should look into just running the appropriate passes, c) the test cases are somewhat constructed. 
To give a concrete example: Something like 
```qsharp
    let qs = (q1, q2);
    CNOT(qs);
```
 would result in the creation of a tuple for qs, followed by storing q1 and q2 in the appropriate items, followed by deconstructing the same tuple again such that CNOT can be invoked.

### CODE REVIEW

**Review by alan-geller — COMMENTED** [review]
A couple of comment typos, otherwise looks great!!

**Review by alan-geller — APPROVED** [review]

File: src/QsCompiler/QirGeneration/Context.cs [hunk 1]
```diff
@@ -813,11 +815,12 @@ bool HasAPredecessor(BasicBlock block)
...
             this.CloseNamingScope();
 
-            if (this.CurrentBlock.Instructions.Count() == 0 && !HasAPredecessor(this.CurrentBlock))
+            if (!HasAPredecessor(this.CurrentBlock)
+                && this.CurrentFunction.BasicBlocks.Count > 1)
             {
                 this.CurrentFunction.BasicBlocks.Remove(this.CurrentBlock);
             }
-            else if (this.CurrentBlock.Terminator == null)
+            else if (this.CurrentBlock.Terminator == null && this.CurrentFunction.ReturnType == this.Context.VoidType)
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply src/QsCompiler/QirGeneration/Context.cs:823]: Yes, this one I am not sure about. The case where the current block is not terminated but the return type is not void does occur, though I am not sure if it is merely a matter of choice of ordering of the blocks (meaning the last/current block is not actually one that can terminate the function) and LLVM is happy with it or not (the validation doesn't complain). 
Unless there is a bug in the handling in one of the control flow statements, then independent on the order in which blocks are listed in the function, the function properly returns a value (or throws) for every terminal block regardless.

[the remark this answers is not in the document]
Comment (alan-geller) [reply src/QsCompiler/QirGeneration/Context.cs:823]: Hmm... Odd. In theory every reachable block should have a terminator. The last block doesn't necessarily have the `return`, for control-flow reasons, but it should have some sort of terminator.
If LLVM doesn't complain, though, I guess we shouldn't either...

File: src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs [hunk 2]
```diff
@@ -607,37 +499,32 @@ IrFunction BuildLiftedSpecialization(string name, QsSpecializationKind kind, ITy
...
+                {
+                    QsResolvedTypeKind.Function paf => paf.Item1,
+                    QsResolvedTypeKind.Operation pao => pao.Item1.Item1,
+                    _ => throw new InvalidOperationException("Partial application of a non-callable value")
+                });
 
-            // And the inputs to the underlying callable
-            var innerTupleType = method.ResolvedType.Resolution switch
+            // Argument type of the callable that is partially applied
+            var innerArgType = method.ResolvedType.Resolution switch
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs]: done.

File: src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestConditional.qs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+namespace Microsoft.Quantum.Testing.QIR{
+
+    newtype CustomTuple = (Data : Double[]);
+
+    function DrawRandom(arg : Double[]) : Int {
+        body intrinsic;
+    }
+
+    function ReturnInt(arg : CustomTuple) : Int {
+
```
Comment (guenp) [inline src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestConditional.qs:13]:
```suggestion
```

File: src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestOpArgument.qs [hunk 4]
```diff
@@ -32,13 +41,23 @@ namespace Microsoft.Quantum.Testing.QIR
...
     @EntryPoint()
-    operation TestOpArgument () : Unit {
-        using (q = Qubit()) {
-            Apply(CNOT(_, q)); 
-            Apply(_SWAP(_, q));
-            Apply(_Choose(q, (_, q))); 
-            Apply(_Choose(_,(q,q)));
+    operation TestOpArgument () : String {
+        using (qs = (Qubit(), Qubit())) {
+
```
Comment (guenp) [inline src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestOpArgument.qs:46]:
```suggestion
```

File: src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestPartials.qs [hunk 5]
```diff
@@ -1,26 +1,51 @@
...
+    }
+
+    operation TakesNestedTuple(tuple : (Int, Double), a1 : String, a2 : Qubit) : Unit is Adj + Ctl {
+    }
+
+    operation ApplyToLittleEndian(bareOp : ((Qubit[]) => Unit is Adj + Ctl), register : LittleEndian)
+    : Unit is Adj + Ctl {
+        bareOp(register!);
+    }
+
```
Comment (guenp) [inline src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestPartials.qs:22]: Not sure if this is the right place for it but maybe we can also test an operation that takes a partial as an input argument? I believe there are some use cases for it (e.g. [ApplyToEach](https://docs.microsoft.com/en-us/qsharp/api/qsharp/microsoft.quantum.canon.applytoeach))

**Review by guenp — APPROVED** [review]
Just some suggestions, otherwise looks great!

File: src/QsCompiler/QirGeneration/Context.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bettinaheim) [reply src/QsCompiler/QirGeneration/Context.cs]: Fixed by properly terminating (inserting unreachable) after a fail.

File: src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestPartials.qs (hunk 5 above)
Comment (bettinaheim) [reply src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestPartials.qs:22]: The tests are not very well structured, sorry. There are tests for that but they are in TestOpArgument.qs.

**Pull request merged**